tests/end-to-end/upgrade/pkg/tests/function/function.go

```diff
@@ -0,0 +1,275 @@
+	"time"
+
+	"github.com/google/uuid"
+	kubelessV1 "github.com/kubeless/kubeless/pkg/apis/kubeless/v1beta1"
+	kubeless "github.com/kubeless/kubeless/pkg/client/clientset/versioned"
+	kymaApi "github.com/kyma-project/kyma/components/api-controller/pkg/apis/gateway.kyma-project.io/v1alpha2"
+	kyma "github.com/kyma-project/kyma/components/api-controller/pkg/clients/gateway.kyma-project.io/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	extensionsv1 "k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	instr "k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+)
+
+// LambdaFunctionUpgradeTest tests the creation of a kubeless function and execute a http request to the exposed api of the function after Kyma upgrade phase
+type LambdaFunctionUpgradeTest struct {
+	functionName, uuid string
+	kubelessClient     kubeless.Interface
+	coreClient         kubernetes.Interface
+	apiClient          kyma.Interface
+	nSpace             string
+	hostName           string
+	stop               <-chan struct{}
+}
+
+func int32Ptr(i int32) *int32 { return &i }
+
+// NewLambdaFunctionUpgradeTest returns new instance of the FunctionUpgradeTest
+func NewLambdaFunctionUpgradeTest(kubelessCli kubeless.Interface, k8sCli kubernetes.Interface, kymaAPI kyma.Interface) *LambdaFunctionUpgradeTest {
+	domainName := os.Getenv("DOMAIN")
+	if len(domainName) == 0 {
+		logrus.Fatal()
+	}
+	nSpace := strings.ToLower("LambdaFunctionUpgradeTest")
+	hostName := fmt.Sprintf("%s-%s.%s", "hello", nSpace, domainName)
+	return &LambdaFunctionUpgradeTest{
+		kubelessClient: kubelessCli,
+		coreClient:     k8sCli,
+		functionName:   "hello",
+		uuid:           uuid.New().String(),
+		nSpace:         nSpace,
+		hostName:       hostName,
+		apiClient:      kymaAPI,
+	}
+}
+
+// CreateResources creates resources needed for e2e upgrade test
+func (f *LambdaFunctionUpgradeTest) CreateResources(stop <-chan struct{}, log logrus.FieldLogger, namespace string) error {
+	log.Println("FunctionUpgradeTest creating resources")
+	f.nSpace = namespace
+	f.stop = stop
+
+	err := f.createFunction()
+	if err != nil {
+		return err
+	}
+
+	err = f.createAPI()
+	if err != nil {
+		log.Printf("create api %v", err)
+		return err
+	}
+
+	// Ensure resources works
+	f.TestResources(stop, log, namespace)
+	if err != nil {
+		log.Printf("First call to TestResources() failed %v", err)
+		return err
+	}
+	return nil
+}
+
+// TestResources tests resources after the upgrade test
+func (f *LambdaFunctionUpgradeTest) TestResources(stop <-chan struct{}, log logrus.FieldLogger, namespace string) error {
+	log.Println("FunctionUpgradeTest testing resources")
+	f.stop = stop
+	err := f.getFunctionPodStatus(10 * time.Minute)
+	if err != nil {
+		return err
+	}
+
+	host := fmt.Sprintf("https://%s", f.hostName)
+
+	value, err := f.getFunctionOutput(host, 1*time.Minute, log)
+	if err != nil {
+		return err
```

Comment:
return errors.Wrap(err, "context") to add context to the error and improve error tracking. same for others